### ISSUE TITLE
spikingreinforce: policy gradient is identically zero — (T,1) x (T,) broadcast collapses the REINFORCE loss

### ISSUE THREAD

**Issue by chrisjz** [SyntheticBrains/nematode#276]
## Summary

`spikingreinforce`'s REINFORCE policy loss is computed on mismatched tensor shapes,
so it broadcasts into an outer product. The result is algebraically
`-mean(log_probs) * mean(advantages)` rather than `-mean(log_probs * advantages)`.

Because both update paths **mean-centre their advantages**, `mean(advantages) ≈ 0`,
and the policy gradient is **numerically zero at the default configuration**. This is
the brain's only gradient path — there is no STDP or eligibility-trace alternative —
so `spikingreinforce` does not learn from its policy-gradient term at all.

**Pre-existing.** This predates the `_policy.py` consolidation (#275); that migration
is byte-identical here and neither introduced nor fixed it. The shape has been present
since at least `d0b0d909` (2026-02-01).

## Mechanism

`state_tensor` is `.unsqueeze(0)`'d to `(1, D)`, so `action_probs` is `(1, n_actions)`
and `Categorical(action_probs)` has `batch_shape (1,)`. Each `log_prob` is therefore
shape `(1,)`, not scalar, and `torch.stack` produces `(T, 1)`:

```python
log_probs = torch.stack(log_probs_list)      # (T, 1)   <- should be (T,)
policy_loss = reinforce_policy_loss(log_probs, advantages)   # advantages is (T,)
```

`(T, 1) * (T,)` broadcasts to `(T, T)` — element `[i][j] = log_probs[i] * advantages[j]`
— and its mean is exactly `mean(log_probs) * mean(advantages)`. Every action's
log-probability is weighted by the *average* advantage instead of its own, which
removes the per-action credit assignment REINFORCE depends on.

## Evidence

Reproducing the real shapes and `spikingreinforce`'s own advantage normalisation
(`(returns - mean) / (std + 1e-8)`), on a small policy network:

```
current (T,1) broadcast : loss=-2.649e-08   grad_norm=0.000e+00
with squeeze(0)         : loss=-5.871e-02   grad_norm=4.935e-01
mean(advantages) = 0.00e+00
```

The gradient norm is **exactly zero**, because `d(loss)/d(log_probs) = -mean(adv)/T`
and `mean(adv)` is zero by construction.

Confirming the identity directly:

```
buggy   -(lp*adv).mean() with lp (T,1): -0.023386
mean(lp)*mean(adv)                    : -0.023386   (equal)
correct -(lp*adv).mean()              : -0.350994
```

`advantage_clip` does not rescue it: it defaults to `0.0` (disabled), and even at the
recommended `2.0–3.0` the clip rarely binds, leaving `mean(adv) ≈ 1.5e-8`. Only
aggressive clipping (`1.0`) shifts the mean enough to produce a gradient — and that
gradient is still the wrong one, since it weights every log-prob by the batch-average
advantage.

The batch path is equally affected: it centres **per episode** and concatenates, so the
concatenated mean is `≈ -1.8e-8` even with unequal episode lengths.

## Affected call sites

Both update paths, which are the only two `backward()` calls in the brain:

- `_perform_intra_episode_update` — `spikingreinforce.py:671` (loss), `:688` (backward)
- `_perform_batch_gradient_update` — `spikingreinforce.py:797` (loss), `:826` (backward)

## Why it went unnoticed

- Nothing errors: broadcasting `(T,1)` against `(T,)` is legal, and `.mean()` returns a
  finite scalar, so the loss looks plausible in logs.
- The clamp `torch.clamp(policy_loss, -10.0, 10.0)` keeps the value in a normal-looking
  range.
- The entropy-regularisation term is computed from **detached** probabilities
  (`spikingreinforce.py:486`), so it contributes no gradient either and cannot mask the
  dead policy gradient by accidentally training the network.
- The existing suite asserts shapes and finiteness, not that the gradient is non-zero or
  that credit assignment is per-action.

## Proposed fix

Squeeze the singleton batch dimension at both scoring sites so each log-probability is
scalar-shaped:

```python
log_prob, _entropy, _probs = categorical_logprob_entropy_from_probs(
    action_probs.squeeze(0),
    int(action_idx),
    device=self.device,
)
```

(Equivalently, drop the `.unsqueeze(0)` on `state_tensor` — but the squeeze at the
scoring site is the smaller change and keeps the network's batched forward intact.)

## This needs validation, not just a patch

The fix **changes training behaviour** — it turns a zero gradient into a real one — so
it should not be landed as a silent one-liner:

- Add a regression test asserting the policy gradient is non-zero for non-degenerate
  advantages, and that `log_probs` is `(T,)` rather than `(T, 1)`.
- Re-run `spikingreinforce` on a foraging config before/after and record the change.
  Any published result involving this brain was produced with a dead policy gradient,
  so its historical numbers describe an untrained policy and should be re-checked
  rather than compared directly.
- Check whether the tuned hyperparameters (`learning_rate`, `entropy_beta`,
  `advantage_clip`) were chosen against the broken gradient and need revisiting.

## Related

- Found while reviewing #275 (`_policy.py` consolidation). That PR deliberately left
  this alone: it is byte-identical here (loss `0.22559338808059692` pre- and
  post-migration), and fixing it there would have broken that change's stated
  no-behavioural-change contract.
- **Checked: no sibling brain shares the pattern.** `spikingreinforce` is the only brain
  that scores a `(1, n)` probability vector (every other scoring site passes a 1-D vector
  or a properly batched `(B, n)` with matching `(B,)` actions), so this is isolated.
